clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java

```diff
@@ -221,7 +222,17 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
      */
     @Override
     public PollResult pollOnClose(long currentTimeMs) {
-        if (membershipManager().isLeavingGroup()) {
+        AbstractMembershipManager<R> membershipManager = membershipManager();
+        GroupMembershipOperation leaveGroupOperation = membershipManager.leaveGroupOperation();
+
+        if (membershipManager.isLeavingGroup() &&
+            // Default operation: both static and dynamic consumers will send a leave heartbeat
+            (GroupMembershipOperation.DEFAULT == leaveGroupOperation ||
+                // Leave operation: both static and dynamic consumers will send a leave heartbeat
+                GroupMembershipOperation.LEAVE_GROUP == leaveGroupOperation ||
+                // Remain in group: only static consumers will send a leave heartbeat, while dynamic members will not
+                membershipManager.groupInstanceId().isPresent())
+        ) {
```

Comment:
Could we let this logic live inside the `MembershipManager.isLeavingGroup()` implementation rather than exposing it here? The `MembershipManager` already manages the group instance ID and `GroupMembershipOperation` values internally. Then we wouldn't have to add the other `AbstractMembershipManager` changes, right?